File: 442-find-all-duplicates-in-an-array/442-find-all-duplicates-in-an-array.cpp
class Solution {
public:
    vector<int> findDuplicates(vector<int>& nums) {
        int n = nums.size();
        vector<int> result;
        
        for( int i = 0; i < n; i++)  {
            while(nums[i] != i + 1) {
                int d = nums[i] - 1;
                if(nums[i] != nums[d])
                    swap(nums[i], nums[d]);
                else
                    break;
            }
        }
        
        for( int i = 0; i < n; i++) {
            if(nums[i] != i + 1)
                result.push_back(nums[i]);
        }
        
        return result;
    }
};

File: 442-find-all-duplicates-in-an-array/442-find-all-duplicates-in-an-array.py
class Solution(object):
    def findDuplicates(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        
        result = []
        
        for i in range(len(nums)):
            while nums[i] != i + 1:
                d = nums[i] - 1
                if nums[i] != nums[d]:
                    nums[i], nums[d] = nums[d], nums[i]
                else:
                    break
        
        for i in range(len(nums)):
            if nums[i] != i + 1:
                result.append(nums[i])
        
        return result
        